### ISSUE TITLE
Training memory usage

### ISSUE THREAD

**Issue by devzhk** [neuralgcm/neuralgcm#281]
Hi, 

Thanks for sharing the great work!
I had a question regarding the GPU memory requirements for training NeuralGCM. While Table G6 reports inference memory usage (which appears to be quite small), I noticed that the batch size per device during training is 1.

Could you provide more details on the GPU memory consumption per training sample? If the memory usage is high, what are the main contributing factors? Specifically, is it primarily due to:
- Backpropagation through the computation graph of the unrolled neural network, or
- Unrolling the dynamical core, which might increase memory overhead?

Any insights into the potential factors and memory optimizations (assuming high memory usage) would be greatly appreciated.

Thanks in advance!

**Comment by shoyer** [comment]
The main reason why we used a batch size of 1 per device (and also model parallelism) is that we found training scaled poorly with batch size -- we saw highly diminishing returns even going from a global batch size of 16 to 32. To obtain the best models, we needed to take as many gradient descent steps as possible. Memory usage was a concern but not the leading one.

I don't have specifics on the GPU memory consumption per training sample (also we trained these models entirely using TPUs) but I recall roughly comparable contributions from neural network, the dycore, and the loss functions.

To make memory usage reasonable, one absolutely essential step is to use [selective gradient checkpointing](https://docs.jax.dev/en/latest/gradient-checkpointing.html) to recompute instead of saving all ouputs from the forward pass. This means memory usage was limited to roughly the cost of saving all outputs required for a single forward time step.
